packages/market-service/src/yearn/yearn.ts

```diff
@@ -0,0 +1,269 @@
+      if (!vault) return null
+
+      if (new BigNumber(vault.underlyingTokenBalance.amountUsdc).eq(0)) {
+        return {
+          price: '0',
+          marketCap: '0',
+          volume: '0',
+          changePercent24Hr: 0
+        }
+      }
+
+      let volume = new BigNumber('0')
+      let changePercent24Hr = 0
+
+      const price = new BigNumber(vault.underlyingTokenBalance.amountUsdc)
+        .div('1e+6')
+        .div(vault.underlyingTokenBalance.amount)
+        .times(`1e+${vault.decimals}`)
+        .times(vault.metadata.pricePerShare)
+        .div(`1e+${vault.decimals}`)
+        .toFixed(2)
+
+      const marketCap = new BigNumber(vault.underlyingTokenBalance.amountUsdc)
+        .div('1e+6')
+        .toFixed(2)
+
+      const historicEarnings = vault.metadata.historicEarnings
+      const lastHistoricalEarnings = historicEarnings
+        ? historicEarnings[historicEarnings.length - 1]
+        : null
+      const secondToLastHistoricalEarnings = historicEarnings
+        ? historicEarnings[historicEarnings.length - 2]
+        : null
+      if (lastHistoricalEarnings && secondToLastHistoricalEarnings) {
+        volume = new BigNumber(lastHistoricalEarnings.earnings.amountUsdc)
+          .minus(secondToLastHistoricalEarnings.earnings.amountUsdc)
+          .div('1e+6')
+          .dp(2)
+      }
+
+      if (lastHistoricalEarnings) {
+        changePercent24Hr =
+          volume
+            .div(lastHistoricalEarnings.earnings.amountUsdc)
+            .times('1e+6')
+            .toNumber() || 0
+      }
+
+      return {
+        price,
+        marketCap,
+        volume: volume.abs().toString(),
+        changePercent24Hr
+      }
+    } catch (e) {
+      console.warn(e)
+      throw new Error('YearnMarketService(findByCaip19): error fetching market data')
+    }
+  }
+
+  private getDate(daysAgo: number) {
+    const date = new Date()
+    date.setDate(date.getDate() - daysAgo)
+    return date
+  }
+
+  findPriceHistoryByCaip19 = async ({
+    caip19,
+    timeframe
+  }: PriceHistoryArgs): Promise<HistoryData[]> => {
+    const { tokenId } = fromCAIP19(caip19)
+    if (!tokenId) return []
+    const checksumAddress = ethers.utils.getAddress(tokenId)
+
+    let daysAgo
+    switch (timeframe) {
+      case HistoryTimeframe.HOUR:
+        daysAgo = 2
+        break
+      case HistoryTimeframe.DAY:
+        daysAgo = 3
+        break
+      case HistoryTimeframe.WEEK:
+        daysAgo = 7
+        break
+      case HistoryTimeframe.MONTH:
+        daysAgo = 30
+        break
+      case HistoryTimeframe.YEAR:
+        daysAgo = 365
+        break
+      case HistoryTimeframe.ALL:
+        daysAgo = 3650
+        break
+      default:
+        daysAgo = 1
+    }
+
+    try {
+      const vaults = await this.yearnSdk.vaults.get([checksumAddress])
+      const decimals = vaults[0].decimals
+
+      const response: VaultDayDataGQLResponse = (await this.yearnSdk.services.subgraph.fetchQuery(
```

Comment:
can fetchQuery take a generic rather than casting? can we remove the cast? it's ok if we can't

Reply:
It expects 0 generic types, so putting it in there was causing a ts issue.